### PR TITLE
[SandboxIR] Add callbacks for instruction insert/remove/move ops

### PR DESCRIPTION
**Pull request by slackito**

### CODE REVIEW

File: llvm/include/llvm/SandboxIR/Context.h [hunk 1]
```diff
@@ -48,6 +61,21 @@ class Context {
   /// Type objects.
   DenseMap<llvm::Type *, std::unique_ptr<Type, TypeDeleter>> LLVMTypeToTypeMap;
 
+  /// Callbacks called when an IR instruction is about to get removed. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<int, RemoveInstrCallback> RemoveInstrCallbacks;
```
Comment (tschuett) [inline llvm/include/llvm/SandboxIR/Context.h:66]: Do you want to keep them as ints or:
```
using ID = int;
```
?

File: llvm/lib/SandboxIR/Context.cpp [hunk 2]
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
...
 }
 
+void Context::runRemoveInstrCallbacks(Instruction *I) {
+  for (const auto &CBEntry : RemoveInstrCallbacks) {
+    CBEntry.second(I);
+  }
+}
+
+void Context::runInsertInstrCallbacks(Instruction *I) {
+  for (auto &CBEntry : InsertInstrCallbacks) {
```
Comment (tschuett) [inline llvm/lib/SandboxIR/Context.cpp:673]: unnecessary braces

File: llvm/lib/SandboxIR/Context.cpp [hunk 3]
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
   return M;
 }
 
+void Context::runRemoveInstrCallbacks(Instruction *I) {
+  for (const auto &CBEntry : RemoveInstrCallbacks) {
```
Comment (tschuett) [inline llvm/lib/SandboxIR/Context.cpp:667]: unnecessary braces

File: llvm/lib/SandboxIR/Context.cpp [hunk 4]
```diff
@@ -660,4 +663,57 @@ Module *Context::createModule(llvm::Module *LLVMM) {
...
+}
+
+void Context::runInsertInstrCallbacks(Instruction *I) {
+  for (auto &CBEntry : InsertInstrCallbacks) {
+    CBEntry.second(I);
+  }
+}
+
+void Context::runMoveInstrCallbacks(Instruction *I, const BBIterator &WhereIt) {
+  for (auto &CBEntry : MoveInstrCallbacks) {
```
Comment (tschuett) [inline llvm/lib/SandboxIR/Context.cpp:679]: unnecessary braces

File: llvm/include/llvm/SandboxIR/Context.h (hunk 1 above)
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h]: Good point, thanks! Switched it to `CallbackID`.

File: llvm/lib/SandboxIR/Context.cpp (hunk 2 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Context.cpp]: Fixed.

File: llvm/lib/SandboxIR/Context.cpp (hunk 3 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Context.cpp]: Fixed.

File: llvm/lib/SandboxIR/Context.cpp (hunk 4 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Context.cpp]: Fixed.

File: llvm/include/llvm/SandboxIR/Context.h [hunk 5]
```diff
@@ -198,6 +233,26 @@ class Context {
 
   /// \Returns the number of values registered with Context.
   size_t getNumValues() const { return LLVMValueToValueMap.size(); }
+
+  /// Register a callback that gets called when a SandboxIR instruction is about
+  /// to be removed from its parent. Note that this will also be called when
+  /// reverting the creation of an instruction.
+  /// \Returns a callback ID for later deregistration.
+  int registerRemoveInstrCallback(RemoveInstrCallback CB);
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/Context.h:241]:
```suggestion
  CallbackID registerRemoveInstrCallback(RemoveInstrCallback CB);
```
ditto below

File: llvm/include/llvm/SandboxIR/Context.h [hunk 6]
```diff
@@ -48,6 +64,21 @@ class Context {
...
+  /// used as IDs for deregistration.
+  DenseMap<CallbackID, InsertInstrCallback> InsertInstrCallbacks;
+  /// Callbacks called when an IR instruction is about to get moved. Keys are
+  /// used as IDs for deregistration.
+  DenseMap<CallbackID, MoveInstrCallback> MoveInstrCallbacks;
+
+  /// A counter used for assigning callback IDs during registration. The same
+  /// counter is used for all kinds of callbacks so we can detect mismatched
+  /// registration/deregistration.
+  static CallbackID NextCallbackID;
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/Context.h:80]: this shouldn't be static?

File: llvm/lib/SandboxIR/Context.cpp [hunk 7]
```diff
@@ -660,4 +663,54 @@ Module *Context::createModule(llvm::Module *LLVMM) {
...
+         "Callback ID not found in RemoveInstrCallbacks during deregistration");
+}
+
+int Context::registerInsertInstrCallback(InsertInstrCallback CB) {
+  CallbackID ID = NextCallbackID++;
+  InsertInstrCallbacks[ID] = CB;
+  return ID;
+}
+void Context::unregisterInsertInstrCallback(CallbackID ID) {
+  [[maybe_unused]] bool erased = InsertInstrCallbacks.erase(ID);
```
Comment (aeubanks) [inline llvm/lib/SandboxIR/Context.cpp:700]:
```suggestion
  [[maybe_unused]] bool Erased = InsertInstrCallbacks.erase(ID);
```

**Review by aeubanks — COMMENTED** [review]
ultra nit: no period at end of commit title

File: llvm/include/llvm/SandboxIR/Context.h [hunk 8]
```diff
@@ -9,18 +9,34 @@
...
+  using RemoveInstrCallback = std::function<void(Instruction *)>;
+  // A InsertInstrCallback receives the instruction about to be created.
+  using InsertInstrCallback = std::function<void(Instruction *)>;
+  // A MoveInstrCallback receives the instruction about to be moved, the
+  // destination BB and an iterator pointing to the insertion position.
+  using MoveInstrCallback =
+      std::function<void(Instruction *, const BBIterator &)>;
+
+  /// An ID for a registered callback. Used for deregistration.
+  using CallbackID = int;
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/Context.h:38]: Why use an ID and not the raw pointer itself as the ID ?

File: llvm/include/llvm/SandboxIR/Context.h (hunk 6 above)
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h:80]: Depends on whether we want callback IDs to be unique across the whole run or just for a single context object. I figured unique-per-run might be good for debugging (you never get two callback registrations with the same ID), but in practice I don't think it will make a big difference, if at all.

File: llvm/include/llvm/SandboxIR/Context.h (hunk 8 above)
Comment (tschuett) [reply llvm/include/llvm/SandboxIR/Context.h:38]: non-determinism

File: llvm/include/llvm/SandboxIR/Context.h (hunk 6 above)
Comment (aeubanks) [reply llvm/include/llvm/SandboxIR/Context.h:80]: I believe people will run multiple LLVM pipelines in parallel in one process, although with different LLVMContexts. so anything static will unintentionally be shared between those threads and be subject to data races

File: llvm/include/llvm/SandboxIR/Context.h (hunk 8 above)
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h:38]: Pointers to the `std::function` objects stored in a map/vector can get invalidated when registering/unregistering more callbacks, right?

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/Context.h:38]: How are we going to get non-determinism ? The ID itself is not printed anywhere, it's just used as a key to remove the callback if needed. If it's the raw pointer of the function it should still work just as well. Or am I missing something?

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/Context.h:38]: > Pointers to the std::function objects stored in a map/vector can get invalidated when registering/unregistering more callbacks, right?

True but we can get around that in multiple ways, one of which is by allocating unique pointers and getting the raw pointer.

What I don't like about the integer IDs is that:
- they are just being used as a key and they don't really convey anything else. They are not even being used for debugging purposes.
- they need the static? counter

Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h:38]: How is another layer of indirection and an extra heap allocation better than keeping an integer around?

Comment (tschuett) [reply llvm/include/llvm/SandboxIR/Context.h:38]: If you use raw pointers as keys in maps, you will get non-determinism.

The context should own the counter.

File: llvm/include/llvm/SandboxIR/Context.h (hunk 6 above)
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h]: Ok, that's a good reason. Made it non-static.

File: llvm/include/llvm/SandboxIR/Context.h (hunk 5 above)
Comment (slackito) [reply llvm/include/llvm/SandboxIR/Context.h]: done

File: llvm/lib/SandboxIR/Context.cpp (hunk 7 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Context.cpp]: done

File: llvm/include/llvm/SandboxIR/Context.h (hunk 8 above)
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/Context.h:38]: > How is another layer of indirection and an extra heap allocation better than keeping an integer around?

Yeah, it's definitely not great, and we should try to make the code that runs the callbacks as fast as possible.

Since @tschuett brought up non-determinism, should we try to make sure that the callbacks are run in the same order as registered? Because we are currently not only not getting that order but since we are iterating over the DenseMap we are also getting a non-deterministic order. Using a MapVector will fix non-determinism, but perhaps we should try to stick to registration order, and try to change it later if it turns out that it's not needed. I think we could use a vector instead of a map, because removal of a callback shouldn't happen too often (and we won't have too many callbacks) so a linear-time search should be fine. Wdyt?

File: llvm/unittests/SandboxIR/SandboxIRTest.cpp [hunk 9]
```diff
@@ -5962,3 +5963,87 @@ TEST_F(SandboxIRTest, CheckClassof) {
...
+          Moved.push_back(std::make_pair(I, &*Where));
+      });
+
+  Ctx.save();
+  auto *NewI = sandboxir::StoreInst::create(Val, Ptr, /*Align=*/std::nullopt,
+                                            Ret->getIterator(), Ctx);
+  EXPECT_THAT(Inserted, testing::ElementsAre(NewI));
+  EXPECT_THAT(Removed, testing::IsEmpty());
+  EXPECT_THAT(Moved, testing::IsEmpty());
+
```
Comment (tmsri) [inline llvm/unittests/SandboxIR/SandboxIRTest.cpp:6010]: A drive-by comment:  How about testing insertBefore and insertAfter?

Comment (slackito) [reply llvm/unittests/SandboxIR/SandboxIRTest.cpp:6020]: Discussed this offline because @vporpo's prototype only called them on create/erase, but called them insert/remove. We only need create/erase/move callbacks for the vectorizer, so we're only adding those for now, and I have renamed them to create/erase callbacks. This leaves space to add proper insert/remove callbacks in the future if we (or some future user of Sandbox IR) needs them.

Thus, no need to test those here as they have no associated callbacks for now.

**Comment by slackito** [comment]
Please take another look. I have also switched callback ids from `int` to `uint64_t` out of concerns (expressed offline) that we could possibly accidentally overflow a 32-bit integer (maybe registering/deregistering callbacks multiple times per region in a very large compile or something like that).

File: llvm/lib/SandboxIR/Context.cpp [hunk 10]
```diff
@@ -660,4 +663,53 @@ Module *Context::createModule(llvm::Module *LLVMM) {
...
+}
+
+void Context::runMoveInstrCallbacks(Instruction *I, const BBIterator &WhereIt) {
+  for (auto &CBEntry : MoveInstrCallbacks)
+    CBEntry.second(I, WhereIt);
+}
+
+Context::CallbackID Context::registerEraseInstrCallback(EraseInstrCallback CB) {
+  CallbackID ID = NextCallbackID++;
+  EraseInstrCallbacks[ID] = CB;
```
Comment (vporpo) [inline llvm/lib/SandboxIR/Context.cpp:683]: How about an assertion on `EraseInstrCallbacks.size()` checking that it's smaller than say 16 ? It could help catch  issues where you keep on registering callbacks by mistake.

**Review by vporpo — APPROVED** [review]
Looks good to me. 
I think it would make sense to add some assertions on the callback map size.

File: llvm/lib/SandboxIR/Context.cpp (hunk 10 above)
Comment (slackito) [reply llvm/lib/SandboxIR/Context.cpp:690]: Done. I'm not super happy with having an arbitrary limit like this not documented anywhere and with a kinda low value, but it will be useful while we're developing the vectorizer and we can always revisit this decision later.
